src/components/calendar_types.rs

```diff
@@ -0,0 +1,358 @@
+            MONTH_NINE_LEAP,
+            MONTH_TEN_LEAP,
+            MONTH_ELEVEN_LEAP,
+            MONTH_TWELVE_LEAP,
+        ];
+
+        if COMMON_MONTH_CODES.contains(month_code) {
+            return Ok(MonthCode(*month_code));
+        }
+
+        match calendar.identifier() {
+            "chinese" | "dangi" if LUNAR_LEAP_MONTHS.contains(month_code) => {
+                Ok(MonthCode(*month_code))
+            }
+            "coptic" | "ethiopic" | "ethiopicaa" if MONTH_THIRTEEN == *month_code => {
+                Ok(MonthCode(*month_code))
+            }
+            "hebrew" if MONTH_FIVE_LEAP == *month_code => Ok(MonthCode(*month_code)),
+            _ => Err(TemporalError::range()
+                .with_message("MonthCode was not valid for the current calendar.")),
+        }
+    }
+
+    pub(crate) fn try_from_partial_date(
+        partial_date: &PartialDate,
+        calendar: &Calendar,
+    ) -> TemporalResult<Self> {
+        match partial_date {
+            PartialDate {
+                month: Some(month),
+                month_code: None,
+                ..
+            } => Self::try_new(&month_to_month_code(*month)?, calendar),
+            PartialDate {
+                month_code: Some(month_code),
+                month: None,
+                ..
+            } => Self::try_new(month_code, calendar),
+            PartialDate {
+                month: Some(month),
+                month_code: Some(month_code),
+                ..
+            } => {
+                are_month_and_month_code_resolvable(*month, month_code)?;
+                Self::try_new(month_code, calendar)
+            }
+            _ => Err(TemporalError::range()
+                .with_message("Month code needed is required to determine date.")),
+        }
+    }
+
+    pub(crate) fn try_from_partial_date_with_fallback(
+        partial: &PartialDate,
+        calendar: &Calendar,
+        fallback: &impl CalendarMethods,
+    ) -> TemporalResult<Self> {
+        match partial {
+            PartialDate {
+                month: Some(month),
+                month_code: None,
+                ..
+            } => Self::try_new(&month_to_month_code(*month)?, calendar),
+            PartialDate {
+                month_code: Some(month_code),
+                month: None,
+                ..
+            } => Self::try_new(month_code, calendar),
+            PartialDate {
+                month: Some(month),
+                month_code: Some(month_code),
+                ..
+            } => {
+                are_month_and_month_code_resolvable(*month, month_code)?;
+                Self::try_new(month_code, calendar)
+            }
+            PartialDate {
+                month: None,
+                month_code: None,
+                ..
+            } => Ok(Self(fallback.month_code()?)),
+        }
+    }
+
+    pub fn as_month_integer(&self) -> TemporalResult<u8> {
+        ascii_four_to_integer(self.0)
+    }
```

Comment:
Not sure I'm understanding what the integer is used for, but if it's the ordinal `month` number - you can't determine that in lunisolar calendars without knowing the year.